Validate dataset_description.json contents

I am preparing example for the

- https://github.com/bids-standard/bids-specification/pull/1861

and upon initial validation receiving NO error although it should be there since I entered (for extra check) a completely unknown value.

```shell
❯ bids-validator-deno --version
bids-validator v1.14.7-dev.0-68-g53fdba69

❯ grep DatasetType dataset_description.json
    "DatasetType": "prlkjasdfoject",
```

```shell
❯ bids-validator-deno $PWD
This dataset appears to be BIDS compatible.

          Summary:                         Available Tasks:        Available Modalities:
          1 Files, 336 B                                                                
          0 - Subjects 1 - Sessions                                                     

	If you have any questions, please post on https://neurostars.org/tags/bids.

❯ bids-validator $PWD

bids-validator@1.14.5
	1: [ERR] Invalid JSON file. The file is not formatted according the schema. (code: 55 - JSON_SCHEMA_VALIDATION_ERROR)
		./dataset_description.json
			Evidence: .DatasetType should be equal to one of the allowed values

	Please visit https://neurostars.org/search?q=JSON_SCHEMA_VALIDATION_ERROR for existing conversations about this issue.

	2: [ERR] There are no subject folders (labeled "sub-*") in the root of this dataset. (code: 45 - SUBJECT_FOLDERS)

	Please visit https://neurostars.org/search?q=SUBJECT_FOLDERS for existing conversations about this issue.

	1: [WARN] The recommended file /README is missing. See Section 03 (Modality agnostic files) of the BIDS specification. (code: 101 - README_FILE_MISSING)

	Please visit https://neurostars.org/search?q=README_FILE_MISSING for existing conversations about this issue.

        Summary:            Available Tasks:        Available Modalities: 
        1 File, 336B                                                      
        0 - Subjects                                                      
        1 - Session                                                       


	If you have any questions, please post on https://neurostars.org/tags/bids.
```

```console
❯ deno run -A src/bids-validator.ts /tmp/empty      
	[ERROR] A dataset_description.json file is required in the root of the dataset (MISSING_DATASET_DESCRIPTION)


		0 more files with the same issue

	Please visit https://neurostars.org/search?q=MISSING_DATASET_DESCRIPTION for existing conversations about this issue.


          Summary:                         Available Tasks:        Available Modalities:
          0 Files, 0 B                                                                  
          0 - Subjects 1 - Sessions                                                     

	If you have any questions, please post on https://neurostars.org/tags/bids.
❯ echo '{}' > /tmp/empty/dataset_description.json
❯ deno run -A src/bids-validator.ts /tmp/empty
	[ERROR] There are no subject directories (labeled "sub-*") in the root of this dataset.
 (SUBJECT_FOLDERS)

		./dataset_description.json
			Evidence: schema.rules.rules.checks.dataset.SubjectFolders

		1 more files with the same issue

	Please visit https://neurostars.org/search?q=SUBJECT_FOLDERS for existing conversations about this issue.

	[WARNING] The BIDSVersion field of 'dataset_description.json' does not match a known release.
The BIDS Schema used for validation may be out of date.
 (UNKNOWN_BIDS_VERSION)

		./dataset_description.json
			Evidence: schema.rules.rules.checks.dataset.UnknownVersion

		1 more files with the same issue

	Please visit https://neurostars.org/search?q=UNKNOWN_BIDS_VERSION for existing conversations about this issue.

	[WARNING] The recommended file /README is missing.
See Section 03 (Modality agnostic files) of the BIDS specification.
 (README_FILE_MISSING)

		./dataset_description.json
			Evidence: schema.rules.rules.checks.hints.ReadmeFileMissing

		1 more files with the same issue

	Please visit https://neurostars.org/search?q=README_FILE_MISSING for existing conversations about this issue.

	[WARNING] The 'Authors' field of 'dataset_description.json' should contain an array of values -
with one author per value.
This was triggered based on the presence of only one author field.
Please ignore if all contributors are already properly listed.
 (TOO_FEW_AUTHORS)

		./dataset_description.json
			Evidence: schema.rules.rules.checks.hints.TooFewAuthors

		1 more files with the same issue

	Please visit https://neurostars.org/search?q=TOO_FEW_AUTHORS for existing conversations about this issue.

	[ERROR] A JSON flle is missing a key listed as required. (JSON_KEY_REQUIRED)

		./dataset_description.json
			Evidence: missing Name as per schema.rules.rules.dataset_metadata.dataset_description

		1 more files with the same issue

	Please visit https://neurostars.org/search?q=JSON_KEY_REQUIRED for existing conversations about this issue.


          Summary:                         Available Tasks:        Available Modalities:
          1 Files, 3 B                                                                  
          0 - Subjects 1 - Sessions                                                     

	If you have any questions, please post on https://neurostars.org/tags/bids.
```

--- a/bids-validator/src/schema/applyRules.ts
+++ b/bids-validator/src/schema/applyRules.ts
@@ -446,75 +446,73 @@ function evalJsonCheck(
   // Count on other JSON rules to use selectors to match the correct files
   if (context.extension === '.json' && sidecarRule) return
 
-  const json = sidecarRule ? context.sidecar : context.json
+  const json: Record<string, any> = sidecarRule ? context.sidecar : context.json
   for (const [key, requirement] of Object.entries(rule.fields)) {
-    const severity = getFieldSeverity(requirement, context)
     // @ts-expect-error
     const metadataDef = schema.objects.metadata[key]
     const keyName: string = metadataDef.name
-    if (severity && severity !== 'ignore' && !(keyName in json)) {
-      if (requirement.issue?.code && requirement.issue?.message) {
-        context.dataset.issues.add({
-          code: requirement.issue.code,
-          subCode: keyName,
-          location: context.path,
-          severity,
-          rule: schemaPath,
-        }, requirement.issue.message)
-      } else {
-        let code
-        if (severity === 'error') {
-          code = sidecarRule ? 'SIDECAR_KEY_REQUIRED' : 'JSON_KEY_REQUIRED'
-        } else if (severity === 'warning') {
-          code = sidecarRule ? 'SIDECAR_KEY_RECOMMENDED' : 'JSON_KEY_RECOMMENDED'
-        }
-        if (code) {
+    const value = json[keyName]
+    if (value === undefined) {
+      const severity = getFieldSeverity(requirement, context)
+      if (severity && severity !== 'ignore') {
+        if (requirement.issue?.code && requirement.issue?.message) {
           context.dataset.issues.add({
-            code,
+            code: requirement.issue.code,
             subCode: keyName,
             location: context.path,
             severity,
-            issueMessage: `missing ${keyName}`,
             rule: schemaPath,
-          })
+          }, requirement.issue.message)
+        } else {
+          let code
+          if (severity === 'error') {
+            code = sidecarRule ? 'SIDECAR_KEY_REQUIRED' : 'JSON_KEY_REQUIRED'
+          } else if (severity === 'warning') {
+            code = sidecarRule ? 'SIDECAR_KEY_RECOMMENDED' : 'JSON_KEY_RECOMMENDED'
+          }
+          if (code) {
+            context.dataset.issues.add({
+              code,
+              subCode: keyName,
+              location: context.path,
+              severity,
+              rule: schemaPath,
+            })
+          }
         }
       }
+
+      /* Regardless of if key is required/recommended/optional, we do no
+       * further valdiation if it is not present in sidecar.
+       */
+      continue
     }
 
-    /* Regardless of if key is required/recommended/optional, we do no
-     * further valdiation if it is not present in sidecar.
-     */
-    if (!(keyName in context.sidecar)) {
-      return
-    }
-
-    let originFileKey = ''
-    let originFile = ''
-    if (keyName in context.sidecarKeyOrigin) {
-      originFile = `${context.sidecarKeyOrigin[keyName]}`
-      originFileKey = `${originFile}:${keyName}`
-    } else {
+    if (sidecarRule && !(keyName in context.sidecarKeyOrigin)) {
       logger.warning(
         `sidecarKeyOrigin map failed to initialize for ${context.path} on key ${keyName}. Validation caching not active for this key.`,
       )
     }
 
-    if (context.dataset.sidecarKeyValidated.has(originFileKey)) {
-      return
+    const location = sidecarRule ? (context.sidecarKeyOrigin[keyName] ?? '') : context.path
+    const affects = sidecarRule ? [context.path] : undefined
+
+    const keyAddress = `${location}:${keyName}`
+
+    if (sidecarRule && context.dataset.sidecarKeyValidated.has(keyAddress)) {
+      continue
     }
 
     const validate = compile(metadataDef)
-    const result = validate(context.sidecar[keyName])
+    const result = validate(value)
     if (result === false) {
-      const messageBase = `Failed for this file.key: ${originFileKey}`
       if (!validate.errors) {
         context.dataset.issues.add({
           code: 'JSON_SCHEMA_VALIDATION_ERROR',
           subCode: keyName,
-          location: context.path,
-          issueMessage: messageBase,
+          location: location,
           rule: schemaPath,
-          affects: [originFile],
+          affects: affects,
         })
       } else {
         for (let error of validate.errors) {
@@ -522,16 +520,16 @@ function evalJsonCheck(
           context.dataset.issues.add({
             code: 'JSON_SCHEMA_VALIDATION_ERROR',
             subCode: keyName,
-            location: context.path,
-            issueMessage: `${messageBase} ${message}`,
+            location: location,
+            issueMessage: message,
             rule: schemaPath,
-            affects: [originFile],
+            affects: affects,
           })
         }
       }
     }
-    if (originFileKey) {
-      context.dataset.sidecarKeyValidated.add(originFileKey)
+    if (sidecarRule && location) {
+      context.dataset.sidecarKeyValidated.add(keyAddress)
     }
   }
 }

--- a/bids-validator/src/tests/local/bids_examples.test.ts
+++ b/bids-validator/src/tests/local/bids_examples.test.ts
@@ -26,7 +26,7 @@ function formatBEIssue(issue: IssueOutput, dsPath: string) {
   errors.push(
     Row.from([
       colors.red(issue.key),
-      issue.files[0].file.name,
+      issue.files[0].file.path,
       issue.files[0].evidence,
     ]),
   )

